tracer/src/Datadog.Trace/Ci/CiEnvironment/GitCommandGitInfoProvider.cs

```diff
@@ -0,0 +1,144 @@
+// <copyright file="GitCommandGitInfoProvider.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+#nullable enable
+
+using System;
+using System.IO;
+using System.Linq;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.Ci.CiEnvironment;
+
+internal sealed class GitCommandGitInfoProvider : GitInfoProvider
+{
+    private GitCommandGitInfoProvider()
+    {
+    }
+
+    public static IGitInfoProvider Instance { get; } = new GitCommandGitInfoProvider();
+
+    private static DateTimeOffset UnixTimeStampToDateTime(long unixTimeStamp)
+    {
+        const long unixEpochTicks = TimeSpan.TicksPerDay * 719162; // 621,355,968,000,000,000
+        return new DateTimeOffset((unixTimeStamp * TimeSpan.TicksPerSecond) + unixEpochTicks, TimeSpan.Zero);
+    }
+
+    protected override bool TryGetFrom(DirectoryInfo gitDirectory, out IGitInfo? gitInfo)
+    {
+        var localGitInfo = new GitInfo
+        {
+            SourceRoot = gitDirectory.Parent?.FullName
+        };
+
+        gitInfo = localGitInfo;
+
+        try
+        {
+            // Ensure we have permissions to read the git directory
+            var safeDirectory = ProcessHelpers.RunCommand(
+                new ProcessHelpers.Command(
+                    cmd: "git",
+                    arguments: $"config --global --add safe.directory {gitDirectory.FullName}",
+                    workingDirectory: gitDirectory.FullName,
+                    useWhereIsIfFileNotFound: true));
+            if (safeDirectory?.ExitCode != 0)
+            {
+                localGitInfo.Errors.Add($"Error setting safe.directory: {safeDirectory?.Error}");
+            }
+
+            // Get the repository URL
+            var repositoryOutput = ProcessHelpers.RunCommand(
+                new ProcessHelpers.Command(
+                    cmd: "git",
+                    arguments: "ls-remote --get-url",
+                    workingDirectory: gitDirectory.FullName,
+                    useWhereIsIfFileNotFound: true));
+            if (repositoryOutput?.ExitCode == 0)
+            {
+                localGitInfo.Repository = repositoryOutput.Output.Trim();
+            }
+            else
+            {
+                localGitInfo.Errors.Add($"Error getting repository URL: {repositoryOutput?.Error}");
+            }
+
+            // Get the branch name
+            var branchOutput = ProcessHelpers.RunCommand(
+                new ProcessHelpers.Command(
+                    cmd: "git",
+                    arguments: "rev-parse --abbrev-ref HEAD",
```

Comment:
This will return a hash if you're in detached head mode (common in CI) I think 🤔 

Maybe you need to run `git symbolic-ref -q HEAD` first to detect if you're in detached head mode (and so can't get the branch, just the hash)?

Reply:
hmmm, you're right, ok, never mind, that was just what I read! 😄